### PR TITLE
KAFKA-17042: The migration docs should remind users to set "broker.id.generation.enable" when adding broker.id

### PR DESCRIPTION
**Pull request by frankvicky**
In the section: Enter Migration Mode on the Brokers

It requires users to add `broker.id`, but this can produce an error: "broker.id must be greater than or equal to -1 and not greater than reserved.broker.max.id". This error is caused by the ZooKeeper broker using a generated broker ID.

As this phase is temporary, a simple solution is to remind users to add `broker.id.generation.enable=false` if the ZooKeeper broker is using a generated broker ID.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: docs/ops.html [hunk 1]
```diff
@@ -3971,7 +3971,10 @@ <h3>Enter Migration Mode on the Brokers</h3>
 
 # KRaft controller quorum configuration
 controller.quorum.voters=3000@localhost:9093
-controller.listener.names=CONTROLLER</code></pre>
+controller.listener.names=CONTROLLER
+
+# If the ZK broker is using a generated broker ID, disable broker ID generation
+broker.id.generation.enable=false</code></pre>
```
Comment (showuon) [inline docs/ops.html]: I think we should add more context here. 

Questions:
1. if we don't add this line: `broker.id.generation.enable=false`, will it always fail?
2. If user needs to enable `broker.id.generation.enable` in their cluster because it works for years, what could they do?

**Review by showuon — COMMENTED** [review]
Thanks for the PR, left some comments.

File: docs/ops.html (hunk 1 above)
Comment (frankvicky) [reply docs/ops.html]: I see, I will figure out how to clarify these two question. 😃

Comment (chia7712) [reply docs/ops.html]: @showuon that is good question. maybe we can describe two solutions for users:

1. `broker.id.generation.enable=false`. in migration, all brokers should have unique id already so it should be fine. 
2. increase `reserved.broker.max.id`. users need to know the max node id in the cluster

Comment (frankvicky) [reply docs/ops.html]: >If we don't add this line: broker.id.generation.enable=false, will it always fail?

If  `broker.id.generation.enable=true` and the user does not set `reserved.broker.max.id` to a value greater than -1, it will definitely fail; conversely, `broker.id.generation.enable=false` will make `reserved.broker.max.id` ineffective, so it will definitely not fail.

>If the user needs to enable broker.id.generation.enable in their cluster because it has been working for years, what could they do?

It is recommended that the user adjusts `reserved.broker.max.id` to a sufficiently large value to facilitate future migrations.

Comment (frankvicky) [reply docs/ops.html]: oh, as @chia7712 said, maybe the first question won't be a concern ?

**Comment by frankvicky** [comment]
Hi @showuon and @chia7712 
I have add some description for `broker.id.generation.enable=false`, PTAL 😃

File: docs/ops.html [hunk 2]
```diff
@@ -3971,7 +3971,11 @@ <h3>Enter Migration Mode on the Brokers</h3>
 
 # KRaft controller quorum configuration
 controller.quorum.voters=3000@localhost:9093
-controller.listener.names=CONTROLLER</code></pre>
+controller.listener.names=CONTROLLER
+
+# This setting requires the broker ID to be specified, but since the ID already exists due to the migration, it can be accepted.
+# If broker.id.generation cannot be disabled for some reason, it is recommended to adjust the reserved.broker.max.id to a known maximum node ID in the cluster.
+broker.id.generation.enable=false</code></pre>
```
Comment (chia7712) [inline docs/ops.html]: maybe we should not add it by default, since not all users could encounter this trouble. We can add `Note` instead.

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for patch

**Comment by frankvicky** [comment]
Hi @chia7712 , I have do some changes based on feedback, PTAL 😸

**Comment by showuon** [comment]
I think the main question I have is, what does this relate to migration? When doing ZK migration, that means, the ZK broker is already up and healthy, before starting migrating to KRaft. And when entering in the `Enter Migration Mode on the Brokers` section, we ask users to add the following 4 configs:

1. controller.quorum.voters
2. controller.listener.names
3. The controller.listener.name should also be added to listener.security.property.map
4. zookeeper.metadata.migration.enable

Thus, I don't think it will cause the error as described in the JIRA, right? I still don't understand in what circumstance this issue will happen. It might need more explanation. Thanks.

**Comment by chia7712** [comment]
> Thus, I don't think it will cause the error as described in the JIRA, right? I still don't understand in what circumstance this issue will happen. It might need more explanation. Thanks.

thanks for this response, and it inspires me to dig-in the root cause. The issue happens when the zk broker is using generated broker id and it is in migrating.

1. if we define the broker id in config file, the broker id is viewed as invalid since it is in the reserved range (when `broker.id.generation.enable=true`)
2. if we don't define the broker id in config file, the broker id will be evaluated according to `meta.properties` [0] . However, we forgot to update `node.id` and so kraft client will use "node.id=-1" [1] to build and it results in error ..


[0] https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaServer.scala#L261
[1] https://github.com/apache/kafka/blob/e55c28c60b0f021b12c93fa04e360ce7a2e0a5ac/core/src/main/scala/kafka/raft/RaftManager.scala#L230

In short, we don't consider the use case of "generated broker id (the broker id is omitted)" + "zk migration". It seems we need two fixes:

1. update docs to explain the possible error when you define the broker id for the zk broker which is using `broker.id.generation.enable`
2. in order to minimize the changes, we can update both `broker.id` and `node.id` together. The side effect is `KafkaConfig#nodeId` will be a mutable variable.

@showuon WDYT?

**Comment by showuon** [comment]
>  1.   update docs to explain the possible error when you define the broker id for the zk broker which is using broker.id.generation.enable

If my understanding is correct, the possible error is the broker id will be ignored and the generated ID might exceed the range of max id, which cause the error reported in JIRA. Is that right?

 > 2.  in order to minimize the changes, we can update both broker.id and node.id together. The side effect is KafkaConfig#nodeId will be a mutable variable.

IMO, this should be a bug. We should take the generated ID into node.id. WDYT?

**Comment by showuon** [comment]
@frankvicky , any thoughts about it? I'd also like to hear since you are the author of this PR.

**Comment by chia7712** [comment]
> If my understanding is correct, the possible error is the broker id will be ignored and the generated ID might exceed the range of max id, which cause the error reported in JIRA. Is that right?

Yep, ignoring the broker id can pass the reserved ids check but it fails on the default node.id. Explicitly defining the "broker.id=generated id" can update node.id to run Kraft components in zk broker but it fails on reserved ids check.

> We should take the generated ID into node.id.

Yes. Specifically, We can define node.id too in updating broker.id manually no matter how we get the broker.id.

**Comment by frankvicky** [comment]
> @frankvicky , any thoughts about it? I'd also like to hear since you are the author of this PR.

Hi @showuon !
The reason why `broker.id.generation.enable` might need to be disabled has been more precisely explained by @chia7712 compared to my previous explanation.

Regarding the second point from @chia7712 , I'm thinking the potential drawbacks of allowing `node.id` to be mutable.

IMHO, could these changes lead to inconsistencies in cluster coordination and communication? Might they also cause any unpredictable behavior?

**Comment by showuon** [comment]
> Yep, ignoring the broker id can pass the reserved ids check but it fails on the default node.id. 

Sorry, I don't understand what you mean here. Enabling the `broker.id.generation.enable`, we'll Ignore the broker ID config. "But it fails on the default node.id" <-- what does it mean?

> Explicitly defining the "broker.id=generated id" can update node.id to run Kraft components in zk broker but it fails on reserved ids check.

"but it fails on reserved ids check" <-- what does it mean? If the ID is successfully generated, it should already be within the reserved range, why do we need another check?

> Regarding the second point from @chia7712 , I'm thinking the potential drawbacks of allowing node.id to be mutable.

Currently, if node.id is not set, but broker.id is set, we'll set node.id=broker.id. So, we can add, if `broker.id.generation.enable` is enabled, and broker.id, node.id not set, we set them. That should fix the problem in this JIRA, right?

> IMHO, could these changes lead to inconsistencies in cluster coordination and communication? Might they also cause any unpredictable behavior?

Could you explain more about the inconsistencies in cluster coordination and communication situation? I might miss something. Thanks.

**Comment by showuon** [comment]
I think my main point is, the JIRA said:
```
it requires users to add "broker.id", but it can produces error "broker.id must be greater than or equal to -1 and not greater than reserved.broker.max.id" too. 
That is caused by the zk broker is using a generated broker id.
```
the root cause of the issue is the generated ID doesn't pass to node.id, which causes this error when startup, is that correct? If so, why cannot we fix the issue by setting node.id = generated ID, if not set? If not, please kindly let me know why this error will appear? Maybe a reproduce step is helpful. Thanks.

**Comment by chia7712** [comment]
> But it fails on the default node.id" <-- what does it mean?
> If the ID is successfully generated, it should already be within the reserved range, why do we need another check?

The story starts with a zk broker having a generated broker id.

Normally, it does not need to define the broker id in the config file. However, that can't work in migration aince Kraft class needs node id and node id is not initialized with (generated) broker id.

If we add the (generated) broker id to config file to initialize both ids, it can't pass the check due to reservation rules.

> the root cause of the issue is the generated ID doesn't pass to node.id, which causes this error when startup, is that correct? 

You are right and that is the root cause.

> ? If so, why cannot we fix the issue by setting node.id = generated ID, if not set?

That can fix the issue. The concern is that the node.id will be a mutable variable rather than a value from configs. That opens a dangerous door that enable everyone to change it in everywhere.

Maybe another solution is we can skip the id value check if it is in zk migration. The side effect is that zk broker needs to define broker id in config file for migration even though generated id is enabled.

**Comment by frankvicky** [comment]
> Could you explain more about the inconsistencies in cluster coordination and communication situation? I might miss something. Thanks.

If I remember correctly, KafkaConfig can be changed externally. If node.id becomes a variable, users might unintentionally or intentionally change node.id, which is clearly not very secure.

**Comment by showuon** [comment]
Thanks for the response. So I think our problem to solve is, when running migration in `Enter Migration Mode on the Brokers` step, the ZK node "MUST" set broker.id, to avoid node.id set to -1 and fail to startup. @frankvicky , could you please focus on this issue to improve the doc? Thanks.

**Comment by chia7712** [comment]
Thanks for @showuon to clarify the root cause. That is a great offline discussion 😃 

The inconsistency between node.id and broker.id (when generated id is enabled) can be addressed in another issue. I will file it later

**Comment by frankvicky** [comment]
Hi @showuon , @chia7712 
I have rewrite the description of `broker.id` to explain why we need this config, PTAL

File: docs/ops.html [hunk 3]
```diff
@@ -3946,6 +3946,7 @@ <h3>Enter Migration Mode on the Brokers</h3>
   </p>
 
   <ul>
+    <li><a href="#brokerconfigs_broker.id">broker.id</a>: In zk mode, if <code>broker.id.generation.enable</code> is enabled (default is enabled), during migration, if <code>broker.id</code> is not set, the migration will fail because <code>node.id</code> will be -1. Both <code>broker.id</code> and <code>node.id</code> must be non-negative integers.</li>
```
Comment (showuon) [inline docs/ops.html]: 1. `In zk mode`: The section is all saying "zk mode" broker, no need to mention it again.
2. `if broker.id.generatio ...., if ....` : too many `if`. 
3. `the migration will fail because <code>node.id</code> will be -1` : we don't need to say this. You can read the whole `Enter Migration Mode on the Brokers` section again, we didn't say, `The controller.listener.name should also be added to listener.security.property.map because it will cause xxx error....`. We just tell users what to do in what condition.

**Comment by frankvicky** [comment]
Hi @showuon 
I have simplified the description  based on feedback, PTAL

**Comment by chia7712** [comment]
> The inconsistency between node.id and broker.id (when generated id is enabled) can be addressed in another issue. I will file it later

https://issues.apache.org/jira/browse/KAFKA-17077

File: docs/ops.html [hunk 4]
```diff
@@ -3946,6 +3946,7 @@ <h3>Enter Migration Mode on the Brokers</h3>
   </p>
 
   <ul>
+    <li><a href="#brokerconfigs_broker.id">broker.id</a>: If <code>broker.id.generation.enable</code> is enabled (default is enabled) and <code>broker.id</code> is not set during migration, it will fail. Ensure <code>broker.id</code> is set to a non-negative integer.</li>
```
Comment (showuon) [inline docs/ops.html]: Looks better!
Maybe we can just say: `Ensure <code>broker.id</code> is set to a non-negative integer even if <code>broker.id.generation.enable</code> is enabled (default is enabled). ` because:
1. we don't need to mention `during migration` since this is a migration section
2. We can combine these 2 sentences.

WDYT?

Comment (frankvicky) [reply docs/ops.html]: Yes, I think your version is more concise, making it easier to read. 😄 
I will adopt it

**Comment by frankvicky** [comment]
Hi @showuon
I have adjust the description based on your feedback, PTAL

**Review by showuon — APPROVED** [review]
LGTM!

File: docs/ops.html [hunk 5]
```diff
@@ -3946,6 +3946,7 @@ <h3>Enter Migration Mode on the Brokers</h3>
   </p>
 
   <ul>
+    <li><a href="#brokerconfigs_broker.id">broker.id</a>: Ensure <code>broker.id</code> is set to a non-negative integer even if <code>broker.id.generation.enable</code> is enabled (default is enabled).</li>
```
Comment (chia7712) [inline docs/ops.html]: That is another issue I mentioned that the broker id is in the reserved range if it was generated. As we encourage users to set broker.id, maybe we should describe the possible error to save users' life 😄

Comment (showuon) [reply docs/ops.html]: Sounds good. Thanks for pointing it out.

Comment (frankvicky) [reply docs/ops.html]: I will do that 🦸🏼

**Comment by frankvicky** [comment]
Hi @chia7712 , @showuon 
I have added a reminder about potential failures, PTAL 🐧

**Review by chia7712 — APPROVED** [review]
LGTM
